[Feat/#23] 연관관계 매핑 구현

jellyyelly opened the pull request:
# 🚀 Pull Request
이슈와 관련된 레이블, 마일스톤, 댓글에 대한 연관관계 매핑 작업을 완료하여 PR 을 요청합니다

## 구현 내용
- 이슈와 레이블 : 다대다 연관관계 매핑
- 마일스톤과 이슈 : 다대일 연관관계 매핑
- 이슈와 댓글 : 댓글을 Aggregate Root 에서 추가하는 연관관계 메서드 추가

## 관련 이슈
close #23

jellyyelly committed:
[fea/#23] chore: 데이터베이스 스키마 변경

- schema.sql
  - 19, 56~57: 마일스톤 PK를 bigint 타입에서 vachar 타입으로 변경, NAME 속성 삭제
  - 48~49: 기본 키에 NOT NULL 조건에 의해 중복되는 제약조건 삭제
  - 68~69: 이슈와 레이블의 중간 테이블은 외래 키의 변경, 삭제에 자동 동기화 되도록 설정

jellyyelly committed:
[feat/#23] feat: 스키마 변경에 따른 마일스톤의 식별자 타입 변경

- 마일스톤의 기존 Long 타입 식별자를 String 식별자로 변경
  - Milestone 클래스에 Persistable 인터페이스 구현
- 그 외 마일스톤 식별자 변경에 영향을 받는 클래스들에도 적용

jellyyelly committed:
[feat/#23] feat: 엔터티 EqualsAndHashCode 추가

- 식별자로만 비교할 수 있도록 구현

jellyyelly committed:
[feat/#23] test: 레이블 제거에 대한 응답 상태 변경에 따른 수정

- LabelControllerTest 149: 기존 204 No Content 에서 200 OK 로 변경에 따른 테스트 수정

jellyyelly committed:
[feat/#23] 이슈의 [열림/닫힘] 상태 필드 추가

- 이슈의 상태 필드는 기본값으로 true

jellyyelly committed:
[feat/#23] feat: 이슈와 레이블의 중간 테이블 구현

- 이슈와 레이블의 다대다 연관관계 매핑을 위한 중간 테이블 구현

jellyyelly committed:
[feat/#23] feat: 이슈 - 레이블 다대다 연관관계 매핑

- 중간 테이블인 IssueLabel 클래스와 다대일 연관관계 매핑

jellyyelly committed:
[feat/#23] feat: 댓글 추가 메서드 구현

- 이슈가 AggregateRoot 이므로 이슈를 통해 댓글을 추가할 수 있도록 구현

jellyyelly committed:
[feat/#23] feat: 이슈 - 마일스톤 다대일 연관관계 매핑

- AggregateReference 를 사용하여 연관관계 표현

jellyyelly committed:
[feat/#23] 서비스 기능 구현

IssueService
- 다수의 레이블을 추가할 수 있다
- 특정 레이블을 이슈에서 제거할 수 있다
- 마일스톤을 지정할 수 있다
- 마일스톤을 제거할 수 있다

CommentService
- 댓글을 추가할 수 있다

jellyyelly committed:
[feat/#23] feat: 이슈 목록 조회 기능 구현

- 이슈 목록은 리스트를 응답하는 별도의 IssueListResponse DTO 를 통해 JSON 으로 응답한다

jellyyelly committed:
[feat/#23] 이슈 연관관계 매핑 테스트 작성
- 이슈 - 레이블, 이슈 - 마일스톤 연관관계 매핑 테스트 코드 작성

jellyyelly committed:
[feat/#23] refactor: 모든 레이블 조회 기능 리팩터링

- @Query 애너테이션 제거 및 메서드 이름 컨벤션 적용

sharpie1330 commented on src/main/java/com/issuetracker/domain/comment/Comment.java:
```diff
@@ -12,6 +13,7 @@
 @Builder
 @NoArgsConstructor
 @AllArgsConstructor
+@EqualsAndHashCode(of = {"id"}, callSuper = false)
```
DB에 저장할 때는 나노초까지 저장되지만, 조회할 때는 나노초가 제외되어 조회되어 callSuper을 false로 한 부분입니다.
혹시 LocalDateTime 대신, Instant나 Timestamp라는 타입을 사용하면 어떨까 궁금해지네요!

sharpie1330 commented on src/main/java/com/issuetracker/domain/issue/Issue.java:
```diff
@@ -28,4 +35,49 @@ public class Issue extends BaseDateTime {
+    private Set<IssueLabel> issueLabels = new HashSet<>();
+
+    @Column("MILESTONE_ID")
+    private AggregateReference<Milestone, String> milestoneRef;
```
단순히 String milestoneId라는 필드를 가지게 구현하실 거라 생각했는데 AggregateReference를 사용하셨군요!
이 클래스는 참조된 애그리거트 루트의 ID만을 가지고 있다는데 그냥 String milestoneId를 기술하는 것과 무슨 차이가 있는지 조금 궁금하네요

jellyyelly replied:
1. AggregateReference 를 사용하면 지정한 애그리거트 루트에 해당하는 객체를 지정하기 때문에 type-safe 합니다.
2. 엔터티의 식별자를 지정했기 때문에 일관성을 보장합니다
3. 식별자의 변수명을 바꾸거나, 타입을 바꿨을 때 변경되는 부분이 지역적입니다
4. 단순히 문자열이 아니라 마일스톤 애그리거트를 참조하고 있다는 의미가 좀 더 명확해집니다.

sharpie1330 commented on src/main/java/com/issuetracker/domain/issue/IssueLabel.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@Builder
+@AllArgsConstructor
```
이곳에는 NoArgsConstructor가 없는데, 이유가 있나요?

jellyyelly replied:
필드가 1개 뿐이어서 그랬는데, 다시 생각해보니 필드가 1개 뿐이라 빌더 패턴을 적용할 이유가 없어 정적 생성자 메서드로 대체했습니다~

sharpie1330 commented on src/main/java/com/issuetracker/domain/issue/IssueLabel.java:
```diff
@@ -0,0 +1,18 @@
+@Getter
+@Builder
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
```
또 여기에 of가 없는 것은 필드가 하나라서가 맞나요?

jellyyelly replied:
네 맞아요~!

sharpie1330 commented on src/main/java/com/issuetracker/domain/issue/IssueService.java:
```diff
@@ -42,4 +47,46 @@ public void edit(Long issueId, IssueUpdateRequest form) {
+
+    public Issue addLabel(Long issueId, Label label) {
+        Issue issue = issueRepository.findById(issueId).orElseThrow(RuntimeException::new);
+        issue.addLabel(label);
```
비즈니스 로직을 도메인에서 처리하는 게 좋습니다 👍👍

sharpie1330 commented on src/main/java/com/issuetracker/domain/issue/IssueService.java:
```diff
@@ -42,4 +47,46 @@ public void edit(Long issueId, IssueUpdateRequest form) {
+        return issueRepository.save(issue);
+    }
+
+    public Issue addLabels(Long issueId, List<Label> labels) {
```
이건 어떤 경우에 사용될까요? 위 메서드와 어떤 차이가 있을까요?

jellyyelly replied:
`addLabel`: 이미 생성된 이슈에 레이블을 추가할 때 (한 번에 하나씩 api 호출을 기대) 사용합니다
`addLabels`: 이슈를 최초로 생성할 때 레이블을 multi-select 로 여러 개를 선택하여 전송할 때 사용합니다

sharpie1330 commented on src/main/java/com/issuetracker/domain/milestone/Milestone.java:
```diff
@@ -14,15 +17,24 @@
     @Column("MILESTONE_ID")
-    private Long id;
-    private String name;
+    private String id;
```
마일스톤이 중복된 이름으로 생성되지 않는다는 걸 오늘 같이 발견했었죠. 마일스톤의 이름만 필요하기 때문에, 이슈에 마일스톤 자체를 필드로 가질 일이 없다는 걸 알게 됐습니다.
DDD에서 다대일 관계는 지원하지 않는 점을 알고, 순환참조와 같은 문제가 발생할 수 있는 점과, 해결 방안으로 RowMapper 등을 사용할 수 있다는 점도 공부했었습니다!
이 리뷰는 필요하지 않지만, 다시 한 번 복기하면 좋을 것 같아 남겨두어요~

sharpie1330 commented on src/main/java/com/issuetracker/domain/milestone/request/MilestoneUpdateRequest.java:
```diff
@@ -13,7 +15,9 @@
-    private String name;
+    @NotBlank
+    @Size(max = 30)
+    private String id;
```
어떤 마일스톤을 수정해야하는지 명시하기 위해 String id를 포함한 점은 알겠지만, controller 사용 시 pathVariable로 id를 전달받을 것 같아 이 DTO 클래스에서 id 부분은 필요 없을 것 같다는 생각이 듭니다!

jellyyelly replied:
마일스톤의 식별자의 타입을 String 으로 바꿔서 수정 대상인 `name` 필드가 `id` 필드로 대체되었습니다.
해당 부분은 수정 대상압니다!

sharpie1330 commented on src/main/resources/schema.sql:
```diff
@@ -66,5 +65,5 @@ ALTER TABLE COMMENT ADD FOREIGN KEY (MEMBER_ID) REFERENCES MEMBER (MEMBER_ID);
-ALTER TABLE ISSUE_LABEL ADD FOREIGN KEY (ISSUE_ID) REFERENCES ISSUE (ISSUE_ID);
-ALTER TABLE ISSUE_LABEL ADD FOREIGN KEY (LABEL_ID) REFERENCES LABEL (LABEL_ID);
\ No newline at end of file
+ALTER TABLE ISSUE_LABEL ADD FOREIGN KEY (ISSUE_ID) REFERENCES ISSUE (ISSUE_ID) ON UPDATE CASCADE ON DELETE CASCADE;
```
이슈가 삭제될 때 라벨도 함께 삭제되도록 Cascade 옵션을 준 게 좋습니다!
이부분에는 고민이 조금 있는데요, cascade 옵션을 주었으니 어차피 삭제될 테니 이슈만 삭제하는 것과
명시적으로 라벨을 삭제하는 로직을 두고 또 이슈를 삭제하는 게 좋을지가 고민이 듭니다. 어떻게 생각하시는지 궁금하네요

sharpie1330 commented on src/test/java/com/issuetracker/domain/label/LabelControllerTest.java:
```diff
@@ -146,6 +146,6 @@ void deletion() throws Exception{
 
         // then
-        result.andExpect(status().isNoContent());
+        result.andExpect(status().isOk());
```
오 수정해주셨네요! 감사합니다 👍

sharpie1330 commented on src/test/java/com/issuetracker/domain/milestone/MilestoneControllerTest.java:
```diff
@@ -135,10 +135,10 @@ void getMilestones() throws Exception {
 
         MilestoneListResponse openedMilestones = MilestoneListResponse.of(
-                List.of(Milestone.builder().isOpen(true).id(1L).name("openedMilestone").build())
+                List.of(Milestone.builder().isOpen(true).id("openedMilestone").build())
```
open 자체가 '열린'이라는 의미를 가지고 있어 openMilestone이 옳은 표현일 것 같긴 합니다...!

jellyyelly replied:
수정했습니다~

sharpie1330 reviewed (approved):
주저리 리뷰 코멘트가 조금 많았는데요, 그냥 제 생각을 적은 거라 너무 신경쓰시지 않으셔도 됩니다. 항상 제가 질문이 좀 많네요...ㅎㅎ 
옐리 정말 수고하셨습니다!

jellyyelly committed:
[feat/#23] feat: 마일스톤 아이디 검증 로직 구현

- @NotBlank, @Size(max = 30) 을 반복적으로 사용하는 부분을 하나의 애너테이션으로 검증하는 검증 애너테이션 추가
- 검증 애너테이션에 대응하는 argument resolver 추가

jellyyelly committed:
[feat/#23] chore: 마일스톤 수정 폼 복구

- 수정 대상인 아이디 복구

jellyyelly committed:
[feat/#23] feat: 마일스톤 mapper 수정

- 마일스톤 테이블의 PK 를 변경하는 SQL 로 수정

jellyyelly committed:
[feat/#23] IssueLabel 생성자 수정

- 빌더 패턴 제거
- 생성자 접근제한 변경
- 정적 생성자 추가

sharpie1330 commented on src/main/java/com/issuetracker/domain/milestone/MilestoneController.java:
```diff
@@ -31,15 +32,16 @@ public ResponseEntity<MilestoneResponse> create(@Valid @RequestBody MilestoneCre
 
     @DeleteMapping("/{milestoneId}")
-    public ResponseEntity<Void> delete(@PathVariable("milestoneId") Long milestoneId) {
+    public ResponseEntity<Void> delete(@PathVariable("milestoneId") @MilestoneId String milestoneId) {
```
PathVariable로 들어오는 값에 대해 검증하시려고 이 어노테이션을 추가하신건가요?
생성 시에만 id 검증이 필요하고, 그 외에는 이미 존재하는 마일스톤에 대한 접근이라 이 부분이 필요가 있는 부분인지 조금 의문이 드네요

jellyyelly replied:
VPC 실습 후, 검증하려면 infra 계층까지 넘어가야 한다는 점(app 서버 -> db 서버)이 비효율적일 것 같아 만들었습니다.
그리고 컨트롤러에서 반복적으로 사용하는데, `@Size`, `@NotBlank` 애너테이션을 메서드마다 덕지덕지 붙이는 것 보다 좀 더 낫지 않나요..? ㅎㅎ

seedspirit replied:
옐리 말씀을 들어보니 애플리케이션 규모가 커지고 코드가 복잡해질수록 이렇게 검증하는게 좋을 수도 있겠다는 생각이 드네용

sharpie1330 replied:
아 이런 부분은 생각하지 못했네요!! 센스 짱입니다 ㅎㅎ

sharpie1330 reviewed (approved):
추가된 부분에 대한 리뷰입니다!

seedspirit commented on src/main/java/com/issuetracker/domain/milestone/argumentresolver/MilestoneIdArgumentResolver.java:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.web.method.support.HandlerMethodArgumentResolver;
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+public class MilestoneIdArgumentResolver implements HandlerMethodArgumentResolver {
```
👍

seedspirit commented on src/test/java/com/issuetracker/domain/issue/IssueTest.java:
```diff
@@ -66,4 +76,163 @@ void create_fail(){
+    }
+
+    @Test
+    @DisplayName("이슈에 지정된 마일스톤을 제거해도 마일스톤 테이블의 마일스톤이 제거되지 않는다")
```
꼼꼼한 테스팅 좋아요우~~

seedspirit commented on src/main/java/com/issuetracker/domain/issue/Issue.java:
```diff
@@ -28,4 +35,43 @@ public class Issue extends BaseDateTime {
+    private AggregateReference<Milestone, String> milestoneRef;
+
+
+    public void addComment(Comment comment) {
+        this.comments.add(comment);
+    }
+
+    public void addLabel(Label label) {
+        IssueLabel ref = IssueLabel.of(label.getId());
+
+        issueLabels.add(ref);
+    }
+
+    public void addLabels(List<Label> labels) {
+        labels.forEach(this::addLabel);
+    }
+
+    public void deleteLabel(Label label) {
+        IssueLabel ref = IssueLabel.of(label.getId());
+
+        issueLabels.remove(ref);
+    }
+
+    public void assignMilestone(Milestone milestone) {
+        milestoneRef = AggregateReference.to(milestone.getId());
+    }
+
+    public void deleteMilestone() {
+        milestoneRef = null;
+    }
 }
```
메서드 명이 직관적이고 코드도 엄청 깔끔하네요. 고민 많이 하신게 느껴집니다

seedspirit reviewed (approved):
고생 많으셨어요!

seedspirit commented:
귯~~
